util/chunk/disk.go

```diff
@@ -0,0 +1,284 @@
+// chunk is as follow:
+//
+// [size of row0 column0], [size of row0 column1], [size of row0 column2]
+// [data of row0 column0], [data of row0 column1], [data of row0 column2]
+// [size of row1 column0], [size of row1 column1], [size of row1 column2]
+// [data of row1 column0], [data of row1 column1], [data of row1 column2]
+//
+// If a column of a row is null, the size of it is -1 and the data is empty.
+type chunkInDisk struct {
+	*Chunk
+	off int64
+	// offsetsOfRows stores the offset of each row.
+	offsetsOfRows []int64
+}
+
+// WriteTo serializes the chunk into the format of chunkInDisk, and
+// writes to w.
+func (chk *chunkInDisk) WriteTo(w io.Writer) (written int64, err error) {
+	var n int64
+	numRows := chk.NumRows()
+	chk.offsetsOfRows = make([]int64, 0, numRows)
+	var format *diskFormatRow
+	for rowIdx := 0; rowIdx < numRows; rowIdx++ {
+		format = convertFromRow(chk.GetRow(rowIdx), format)
+		chk.offsetsOfRows = append(chk.offsetsOfRows, chk.off+written)
+
+		n, err = chk.writeRowTo(w, format)
+		written += n
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+// writeRowTo serializes a row of the chunk into the format of
+// diskFormatRow, and writes to w.
+func (chk *chunkInDisk) writeRowTo(w io.Writer, format *diskFormatRow) (written int64, err error) {
+	n, err := w.Write(i64SliceToBytes(format.sizesOfColumns))
+	written += int64(n)
+	if err != nil {
+		return
+	}
+	for _, data := range format.cells {
+		n, err = w.Write(data)
+		written += int64(n)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+func (chk *chunkInDisk) getOffsetsOfRows() []int64 { return chk.offsetsOfRows }
+
+// rowInDisk represents a Row in format of diskFormatRow.
+type rowInDisk struct {
+	numCol int
+	diskFormatRow
+}
+
+// ReadFrom reads data of r, deserializes it from the format of diskFormatRow
+// into Row.
+func (row *rowInDisk) ReadFrom(r io.Reader) (n int64, err error) {
+	b := make([]byte, 8*row.numCol)
+	var n1 int
+	n1, err = io.ReadFull(r, b)
+	n += int64(n1)
+	if err != nil {
+		return
+	}
+	row.sizesOfColumns = bytesToI64Slice(b)
+	row.cells = make([][]byte, 0, row.numCol)
+	for _, size := range row.sizesOfColumns {
+		if size == -1 {
+			continue
+		}
+		cell := make([]byte, size)
+		row.cells = append(row.cells, cell)
+		n1, err = io.ReadFull(r, cell)
+		n += int64(n1)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+// diskFormatRow represents a row in a chunk in disk format. The disk format
+// of a row is described in the doc of chunkInDisk.
+type diskFormatRow struct {
+	// sizesOfColumns stores the size of each column in a row.
+	// -1 means the value of this column is null.
+	sizesOfColumns []int64 // -1 means null
+	cells          [][]byte
+}
+
+// convertFromRow serializes one row of chunk to diskFormatRow
+func convertFromRow(row Row, reuse *diskFormatRow) (format *diskFormatRow) {
+	numCols := row.Chunk().NumCols()
+	if reuse != nil {
+		format = reuse
```

Comment:
s/ format/ rowInDiskFormat ?

Reply:
I think it's better to keep the short-lived value short.